cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -171,14 +183,32 @@ public final RuntimeInfo run(ProgramDescriptor programDescriptor, ProgramOptions
         // Get the artifact details and save it into the program options.
         ArtifactId artifactId = programDescriptor.getArtifactId();
         ArtifactDetail artifactDetail = getArtifactDetail(artifactId);
+        ApplicationSpecification appSpec = programDescriptor.getApplicationSpecification();
+        ProgramDescriptor newProgramDescriptor = programDescriptor;
+
+        boolean isPreview = Boolean.valueOf(
+          options.getArguments().getOption(ProgramOptionConstants.IS_PREVIEW, "false"));
+        // do the app spec regeneration if the mode is on premise, for isolated mode, the regeneration is done on the
+        // runtime environment before the program launch
+        // for preview we already have a resolved app spec, so no need to regenerate the app spec again
+        if (!isPreview && appSpec != null && ClusterMode.ON_PREMISE.equals(clusterMode)) {
+          try {
+            ApplicationSpecification generatedAppSpec =
+              regenerateAppSpec(artifactDetail, programId, artifactId, appSpec, options);
+            appSpec = generatedAppSpec != null ? generatedAppSpec : appSpec;
+            newProgramDescriptor = new ProgramDescriptor(programDescriptor.getProgramId(), appSpec);
+          } catch (Exception e) {
```

Comment:
Is it possible to catch specific exceptions here?

Reply:
This is to preserve backward compatibility, if anything bad happens, we should fall back to use old app spec